elastic/kibana
# [dashboard] decouple "link to library" and "unlink from library" actions from AttributeService

There are 2 paths for decoupling from AttributeService
1. replace existing actions with new actions that use `HasLibraryTransforms` interface. Update all embeddables to implement `HasLibraryTransforms` interfaces.
2. Rename existing actions as `legacy`. Create new actions that use `HasLibraryTransforms` interface. Provide a reference implementation for new actions. Convert embeddables to new `HasLibraryTransforms` interface as the embeddables get converted to react embeddables.

Option 2 was chosen to limit scope and gradually convert embeddables to `HasLibraryTransforms` interface.

This PR:
1. Remove `@kbn/presentation-library` package. Interfaces in this package have been rolled into `@kbn/presentation-publishing` package.
2. Rename existing interface as `HasLegacyLibraryTransforms`
3. Create new interface `HasLibraryTransforms` 
4. Rename `AddToLibraryAction` action to `LegacyAddToLibraryAction`. Modify action to use `HasLegacyLibraryTransforms` interface and guards.
5. Rename `UnlinkFromLibraryAction`  action to `LegacyUnlinkFromLibraryAction`. Modify action to use `HasLegacyLibraryTransforms` interface and guards.
6. Rename `LibraryNotificationAction` action to `LegacyLibraryNotificationAction`. Modify action to use `LegacyUnlinkFromLibraryAction`.
7. Create new `AddToLibraryAction`. Code action to use `HasLibraryTransforms` interface.
8. Create new `UnlinkFromLibraryAction`. Code action to use `HasLibraryTransforms` interface.
9. Create new `LibraryNotificationAction`. Code action to use `UnlinkFromLibraryAction` action.
10. Update MapEmbeddable to implement `HasLibraryTransforms` interface so that Map embeddable can be used to test new actions.

### Test
1. install sample web logs
2. create new dashboard
3. Click "Add panel" and select "Maps".
4. Click "Save and return".
5. Save dashboard. Inspect dashboard saved object. Verify panel is by-value and contains `attributes` in `panelsJSON`
    <img width="300" alt="Screenshot 2024-03-22 at 2 49 56 PM" src="https://github.com/elastic/kibana/assets/373691/49189613-f7c4-435d-88ab-d9c8ceb1575f">
6. Go back to dashboard and open context menu. Click "more" and then click "Save to library".
7. Save dashboard. Inspect dashboard saved object. Verify panel is by-reference and does not contain `attributes` in `panelsJSON`.
    <img width="300" alt="Screenshot 2024-03-22 at 2 52 19 PM" src="https://github.com/elastic/kibana/assets/373691/e3b2eace-a48d-4dd0-a771-f22436d72935">
8. Go to maps listing page. Verify map is displayed in listing page. Open map and verify it opens. Add some new layers and save map.
11. Go back to dashboard. Verify map contains new layers added to saved object.
12. Open context menu. Click "more" and then click "Unlink from library".
13. Save dashboard and verify map panel is now by-value again.
